File: src/systems/lod/cbt/cbt_classify.ts
import { Matrix, Plane, Vector3 } from '@babylonjs/core';
import type { CbtNode } from './cbt_state';

export type CbtSplitCandidate = {
    nodeId: number;
    score: number;
    projectedAreaPx2: number;
};

export type CbtLeafMetric = {
    nodeId: number;
    parentId: number | null;
    projectedAreaPx2: number;
};

export type CbtClassifyParams = {
    leaves: ReadonlyArray<CbtNode>;
    cameraWorldDouble: Vector3;
    planetCenterWorldDouble: Vector3;
    renderParentWorldMatrix: Matrix;
    viewportHeightPx: number;
    cameraFovRadians: number;
    splitThresholdPx2: number;
    splitHysteresis: number;
    /**
     * When true, leaves on the planet's far hemisphere are excluded from split
     * candidates (paper-style backside cull). Merges are still allowed for
     * culled leaves so detail is reclaimed off-screen. Default false.
     */
    cullBackface?: boolean;
    /**
     * Guard band for the backside cull, as a cosine threshold on
     * dot(surfaceNormal, dirToCamera). 0 = exact horizon plane; a small negative
     * value (default -0.05) keeps a thin band just behind the horizon to avoid
     * silhouette pop when rotating. Only used when {@link cullBackface} is true.
     */
    cullMinDot?: number;
    /**
     * Render-space camera frustum planes (from `Frustum.GetPlanes`). When set,
     * leaves whose bounding sphere is fully outside the frustum (beyond the guard
     * band) are excluded from split candidates — off-screen geometry stops
     * refining. Merges still see every leaf, so off-screen detail collapses.
     */
    frustumPlanes?: ReadonlyArray<Plane> | null;
    /**
     * Frustum guard band as a multiple of each triangle's bounding radius. The
     * effective cull margin is `boundRadius * (1 + guard)`, so a band around the
     * view keeps refining (prefetch) to avoid pop when the camera turns. Default 1.
     */
    frustumGuardScale?: number;
};

const MIN_DISTANCE = 1.0;
const DEFAULT_CULL_MIN_DOT = -0.05;
const DEFAULT_FRUSTUM_GUARD = 1.0;

/**
 * Is a triangle's bounding sphere fully outside the frustum (beyond the margin)?
 * `r*` is the sphere centre in render space; planes have inward normals.
 */
function outsideFrustum(
    rx: number, ry: number, rz: number,
    boundRadius: number,
    planes: ReadonlyArray<Plane>,
    guardScale: number
): boolean {
    const margin = boundRadius * (1 + guardScale);
    for (let i = 0; i < planes.length; i++) {
        const pl = planes[i];
        const d = pl.normal.x * rx + pl.normal.y * ry + pl.normal.z * rz + pl.d;
        if (d < -margin) return true; // fully on the outside of this plane
    }
    return false;
}

/**
 * Backside test for a leaf centroid. `r*` is the outward surface normal
 * (centroid − planetCenter); `t*` is the direction to the camera
 * (camera − centroid). Returns true when the leaf faces away from the camera
 * beyond the guard band. One sqrt; no allocation.
 */
function isBackface(
    rx: number, ry: number, rz: number,
    tx: number, ty: number, tz: number,
    minDot: number
): boolean {
    const dot = rx * tx + ry * ty + rz * tz;
    if (minDot === 0) return dot <= 0;
    const rl2 = rx * rx + ry * ry + rz * rz;
    const tl2 = tx * tx + ty * ty + tz * tz;
    if (rl2 < 1e-24 || tl2 < 1e-24) return false;
    return dot < minDot * Math.sqrt(rl2 * tl2);
}

/**
 * Is a single vertex behind the camera's horizon? `v` is the leaf vertex in
 * planet-local space; `matrix` is the planet's render-parent world matrix (its
 * rotation maps local → world-relative-to-planet-center); `camRel*` is the
 * camera position relative to the planet center. Writes into `tmp`.
 */
function vertexBackface(
    v: Vector3,
    matrix: Matrix,
    camRelX: number, camRelY: number, camRelZ: number,
    minDot: number,
    tmp: Vector3
): boolean {
    Vector3.TransformNormalToRef(v, matrix, tmp);
    const rx = tmp.x, ry = tmp.y, rz = tmp.z; // radial = vertexWorld − planetCenter
    return isBackface(rx, ry, rz, camRelX - rx, camRelY - ry, camRelZ - rz, minDot);
}

/**
 * A triangle is back-facing (safe to cull from split candidates) only when ALL
 * THREE vertices are behind the horizon. Testing the centroid alone wrongly culls
 * huge coarse triangles that straddle the horizon — which then can never split to
 * produce front-facing children (the startup "stuck at minimum LOD" bug).
 */
function triangleBackface(
    v0: Vector3, v1: Vector3, v2: Vector3,
    matrix: Matrix,
    camRelX: number, camRelY: number, camRelZ: number,
    minDot: number,
    tmp: Vector3
): boolean {
    return (
        vertexBackface(v0, matrix, camRelX, camRelY, camRelZ, minDot, tmp) &&
        vertexBackface(v1, matrix, camRelX, camRelY, camRelZ, minDot, tmp) &&
        vertexBackface(v2, matrix, camRelX, camRelY, camRelZ, minDot, tmp)
    );
}

function triangleArea(v0: Vector3, v1: Vector3, v2: Vector3): number {
    // Inlined cross-product magnitude — no Vector3 allocations on this hot path.
    // Same component order as Vector3.Cross(v1-v0, v2-v0).length() so results are
    // bit-identical to the previous implementation (golden hash preserved).
    const e0x = v1.x - v0.x;
    const e0y = v1.y - v0.y;
    const e0z = v1.z - v0.z;
    const e1x = v2.x - v0.x;
    const e1y = v2.y - v0.y;
    const e1z = v2.z - v0.z;
    const cx = e0y * e1z - e0z * e1y;
    const cy = e0z * e1x - e0x * e1z;
    const cz = e0x * e1y - e0y * e1x;
    return Math.sqrt(cx * cx + cy * cy + cz * cz) * 0.5;
}

export function classifySplitCandidates({
    leaves,
    cameraWorldDouble,
    planetCenterWorldDouble,
    renderParentWorldMatrix,
    viewportHeightPx,
    cameraFovRadians,
    splitThresholdPx2,
    splitHysteresis,
}: CbtClassifyParams): CbtSplitCandidate[] {
    const focal = viewportHeightPx / (2 * Math.tan(cameraFovRadians * 0.5));
    const threshold = splitThresholdPx2 * Math.max(0.05, Math.min(1.0, splitHysteresis));

    const tmpCentroidLocal = new Vector3();
    const tmpCentroidRotated = new Vector3();
    const tmpCentroidWorld = new Vector3();

    const candidates: CbtSplitCandidate[] = [];

    for (const leaf of leaves) {
        tmpCentroidLocal
            .copyFrom(leaf.v0)
            .addInPlace(leaf.v1)
            .addInPlace(leaf.v2)
            .scaleInPlace(1 / 3);

        Vector3.TransformNormalToRef(tmpCentroidLocal, renderParentWorldMatrix, tmpCentroidRotated);
        tmpCentroidWorld.copyFrom(planetCenterWorldDouble).addInPlace(tmpCentroidRotated);

        const distance = Math.max(MIN_DISTANCE, Vector3.Distance(cameraWorldDouble, tmpCentroidWorld));
        const areaWorld = triangleArea(leaf.v0, leaf.v1, leaf.v2);
        const projectedAreaPx2 = areaWorld * (focal * focal) / (distance * distance);

        if (projectedAreaPx2 >= threshold) {
            candidates.push({
                nodeId: leaf.id,
                score: projectedAreaPx2,
                projectedAreaPx2,
            });
        }
    }

    candidates.sort((a, b) => b.score - a.score);
    return candidates;
}

export type CbtClassifyResult = {
    /** Leaves whose projected area exceeds the split threshold, sorted desc. */
    splitCandidates: CbtSplitCandidate[];
    /** Parent ids eligible to merge (both children present, below threshold), sorted asc by area. */
    mergeParents: number[];
};

/**
 * Single-pass classification: computes each leaf's projected area exactly once
 * and derives BOTH the split candidates and the merge-parent list. This is the
 * behavior-preserving merge of {@link classifySplitCandidates} and the
 * scheduler's merge aggregation that previously ran as two O(n) passes.
 *
 * Thresholds replicate the prior code exactly:
 *  - split:  area >= splitThresholdPx2 · clamp(hysteresis, 0.05, 1)
 *  - merge:  parent has 2 leaf children AND maxChildArea <= splitThresholdPx2 · hysteresis
 */
export function classifyLeaves({
    leaves,
    cameraWorldDouble,
    planetCenterWorldDouble,
    renderParentWorldMatrix,
    viewportHeightPx,
    cameraFovRadians,
    splitThresholdPx2,
    splitHysteresis,
    cullBackface = false,
    cullMinDot = DEFAULT_CULL_MIN_DOT,
    frustumPlanes = null,
    frustumGuardScale = DEFAULT_FRUSTUM_GUARD,
}: CbtClassifyParams): CbtClassifyResult {
    const focal = viewportHeightPx / (2 * Math.tan(cameraFovRadians * 0.5));
    const splitThreshold = splitThresholdPx2 * Math.max(0.05, Math.min(1.0, splitHysteresis));
    const mergeThreshold = splitThresholdPx2 * splitHysteresis;

    const tmpCentroidLocal = new Vector3();
    const tmpCentroidRotated = new Vector3();
    const tmpCentroidWorld = new Vector3();
    const tmpVertex = new Vector3();

    // Camera position relative to the planet centre (for the horizon cull).
    const camRelX = cameraWorldDouble.x - planetCenterWorldDouble.x;
    const camRelY = cameraWorldDouble.y - planetCenterWorldDouble.y;
    const camRelZ = cameraWorldDouble.z - planetCenterWorldDouble.z;

    const splitCandidates: CbtSplitCandidate[] = [];
    const parentAgg = new Map<number, { children: number; maxAreaPx2: number }>();

    for (const leaf of leaves) {
        tmpCentroidLocal
            .copyFrom(leaf.v0)
            .addInPlace(leaf.v1)
            .addInPlace(leaf.v2)
            .scaleInPlace(1 / 3);

        Vector3.TransformNormalToRef(tmpCentroidLocal, renderParentWorldMatrix, tmpCentroidRotated);
        tmpCentroidWorld.copyFrom(planetCenterWorldDouble).addInPlace(tmpCentroidRotated);

        const distance = Math.max(MIN_DISTANCE, Vector3.Distance(cameraWorldDouble, tmpCentroidWorld));
        const areaWorld = triangleArea(leaf.v0, leaf.v1, leaf.v2);
        const projectedAreaPx2 = areaWorld * (focal * focal) / (distance * distance);

        if (projectedAreaPx2 >= splitThreshold) {
            // Backside cull applies to split candidates only — merges below still
            // see every leaf. Cull only when ALL THREE vertices are behind the
            // horizon, so big coarse triangles straddling the horizon still split.
            let culled =
                cullBackface &&
                triangleBackface(
                    leaf.v0, leaf.v1, leaf.v2,
                    renderParentWorldMatrix,
                    camRelX, camRelY, camRelZ,
                    cullMinDot,
                    tmpVertex
                );

            // Frustum cull: drop candidates whose bounding sphere is fully outside
            // the view (plus a guard band so a margin keeps refining for prefetch).
            if (!culled && frustumPlanes) {
                let br2 = 0;
                const c = tmpCentroidLocal;
                for (const v of [leaf.v0, leaf.v1, leaf.v2]) {
                    const dx = v.x - c.x, dy = v.y - c.y, dz = v.z - c.z;
                    const d2 = dx * dx + dy * dy + dz * dz;
                    if (d2 > br2) br2 = d2;
                }
                const boundRadius = Math.sqrt(br2);
                culled = outsideFrustum(
                    tmpCentroidWorld.x - cameraWorldDouble.x,
                    tmpCentroidWorld.y - cameraWorldDouble.y,
                    tmpCentroidWorld.z - cameraWorldDouble.z,
                    boundRadius,
                    frustumPlanes,
                    frustumGuardScale
                );
            }

            if (!culled) {
                splitCandidates.push({
                    nodeId: leaf.id,
                    score: projectedAreaPx2,
                    projectedAreaPx2,
                });
            }
        }

        if (leaf.parentId !== null) {
            const prev = parentAgg.get(leaf.parentId) ?? { children: 0, maxAreaPx2: 0 };
            prev.children++;
            prev.maxAreaPx2 = Math.max(prev.maxAreaPx2, projectedAreaPx2);
            parentAgg.set(leaf.parentId, prev);
        }
    }

    splitCandidates.sort((a, b) => b.score - a.score);

    const mergeParents = Array.from(parentAgg.entries())
        .filter(([, agg]) => agg.children === 2 && agg.maxAreaPx2 <= mergeThreshold)
        .sort((a, b) => a[1].maxAreaPx2 - b[1].maxAreaPx2)
        .map(([parentId]) => parentId);

    return { splitCandidates, mergeParents };
}

export function measureLeafProjectedAreas({
    leaves,
    cameraWorldDouble,
    planetCenterWorldDouble,
    renderParentWorldMatrix,
    viewportHeightPx,
    cameraFovRadians,
}: Omit<CbtClassifyParams, 'splitThresholdPx2' | 'splitHysteresis'>): CbtLeafMetric[] {
    const focal = viewportHeightPx / (2 * Math.tan(cameraFovRadians * 0.5));
    const tmpCentroidLocal = new Vector3();
    const tmpCentroidRotated = new Vector3();
    const tmpCentroidWorld = new Vector3();

    const metrics: CbtLeafMetric[] = [];
    for (const leaf of leaves) {
        tmpCentroidLocal
            .copyFrom(leaf.v0)
            .addInPlace(leaf.v1)
            .addInPlace(leaf.v2)
            .scaleInPlace(1 / 3);

        Vector3.TransformNormalToRef(tmpCentroidLocal, renderParentWorldMatrix, tmpCentroidRotated);
        tmpCentroidWorld.copyFrom(planetCenterWorldDouble).addInPlace(tmpCentroidRotated);

        const distance = Math.max(MIN_DISTANCE, Vector3.Distance(cameraWorldDouble, tmpCentroidWorld));
        const areaWorld = triangleArea(leaf.v0, leaf.v1, leaf.v2);
        const projectedAreaPx2 = areaWorld * (focal * focal) / (distance * distance);
        metrics.push({
            nodeId: leaf.id,
            parentId: leaf.parentId,
            projectedAreaPx2,
        });
    }

    return metrics;
}

File: terrain/src/cbt/cbt_classify.rs
//! Bit-exact Rust port of CBT classification (src/systems/lod/cbt/cbt_classify.ts):
//! projected triangle area in px², backside cull (all 3 vertices), and frustum cull
//! (bounding sphere vs render-space planes). Produces the split-candidate list
//! (sorted desc by area) and merge-parent list (sorted asc by area).
//!
//! IMPORTANT: `focal` is passed IN (computed once on the main thread as
//! `viewportHeightPx / (2*tan(fov/2))`), NOT recomputed here — `tan` is the only
//! transcendental in this path and libm/JS may differ by an ULP, which could flip a
//! split decision at a threshold boundary and diverge the topology. Everything else
//! here is `+`/`*`/`sqrt`, so given the same focal the result is bit-identical to TS.

use crate::cbt::cbt_state::CbtLeaf;

const MIN_DISTANCE: f64 = 1.0;

/// Camera/view parameters for one classification pass (render matrix is the
/// planet render-parent world matrix, Babylon row-major `Matrix.m`).
pub struct ClassifyParams<'a> {
    pub camera_world: [f64; 3],
    pub planet_center_world: [f64; 3],
    pub render_mat: [f64; 16],
    pub focal: f64,
    pub split_threshold_px2: f64,
    pub split_hysteresis: f64,
    pub cull_backface: bool,
    pub cull_min_dot: f64,
    /// Render-space frustum planes as (nx,ny,nz,d); None disables the frustum cull.
    pub frustum_planes: Option<&'a [[f64; 4]]>,
    pub frustum_guard_scale: f64,
}

pub struct ClassifyResult {
    /// Leaf ids whose area exceeds the split threshold (and pass culls), desc by area.
    pub split_candidates: Vec<usize>,
    /// Parent ids eligible to merge (2 leaf children, below threshold), asc by area.
    pub merge_parents: Vec<usize>,
}

/// Babylon Vector3.TransformNormal: upper-left 3x3 of the row-major matrix.
#[inline]
fn transform_normal(c: [f64; 3], m: &[f64; 16]) -> [f64; 3] {
    [
        c[0] * m[0] + c[1] * m[4] + c[2] * m[8],
        c[0] * m[1] + c[1] * m[5] + c[2] * m[9],
        c[0] * m[2] + c[1] * m[6] + c[2] * m[10],
    ]
}

/// Inlined cross-product magnitude * 0.5 — same component order as the TS
/// `triangleArea` (load-bearing for matching projected areas exactly).
#[inline]
fn triangle_area(v0: [f64; 3], v1: [f64; 3], v2: [f64; 3]) -> f64 {
    let e0x = v1[0] - v0[0];
    let e0y = v1[1] - v0[1];
    let e0z = v1[2] - v0[2];
    let e1x = v2[0] - v0[0];
    let e1y = v2[1] - v0[1];
    let e1z = v2[2] - v0[2];
    let cx = e0y * e1z - e0z * e1y;
    let cy = e0z * e1x - e0x * e1z;
    let cz = e0x * e1y - e0y * e1x;
    (cx * cx + cy * cy + cz * cz).sqrt() * 0.5
}

#[inline]
fn distance(a: [f64; 3], b: [f64; 3]) -> f64 {
    let dx = a[0] - b[0];
    let dy = a[1] - b[1];
    let dz = a[2] - b[2];
    (dx * dx + dy * dy + dz * dz).sqrt()
}

fn is_backface(r: [f64; 3], t: [f64; 3], min_dot: f64) -> bool {
    let dot = r[0] * t[0] + r[1] * t[1] + r[2] * t[2];
    if min_dot == 0.0 {
        return dot <= 0.0;
    }
    let rl2 = r[0] * r[0] + r[1] * r[1] + r[2] * r[2];
    let tl2 = t[0] * t[0] + t[1] * t[1] + t[2] * t[2];
    if rl2 < 1e-24 || tl2 < 1e-24 {
        return false;
    }
    dot < min_dot * (rl2 * tl2).sqrt()
}

fn vertex_backface(v: [f64; 3], m: &[f64; 16], cam_rel: [f64; 3], min_dot: f64) -> bool {
    let r = transform_normal(v, m); // radial = vertexWorld - planetCenter
    let t = [cam_rel[0] - r[0], cam_rel[1] - r[1], cam_rel[2] - r[2]];
    is_backface(r, t, min_dot)
}

fn triangle_backface(
    v0: [f64; 3],
    v1: [f64; 3],
    v2: [f64; 3],
    m: &[f64; 16],
    cam_rel: [f64; 3],
    min_dot: f64,
) -> bool {
    vertex_backface(v0, m, cam_rel, min_dot)
        && vertex_backface(v1, m, cam_rel, min_dot)
        && vertex_backface(v2, m, cam_rel, min_dot)
}

fn outside_frustum(
    r: [f64; 3],
    bound_radius: f64,
    planes: &[[f64; 4]],
    guard_scale: f64,
) -> bool {
    let margin = bound_radius * (1.0 + guard_scale);
    for pl in planes {
        let d = pl[0] * r[0] + pl[1] * r[1] + pl[2] * r[2] + pl[3];
        if d < -margin {
            return true;
        }
    }
    false
}

pub fn classify_leaves(leaves: &[CbtLeaf], p: &ClassifyParams) -> ClassifyResult {
    let split_threshold = p.split_threshold_px2 * p.split_hysteresis.clamp(0.05, 1.0);
    let merge_threshold = p.split_threshold_px2 * p.split_hysteresis;
    let focal2 = p.focal * p.focal;

    let cam_rel = [
        p.camera_world[0] - p.planet_center_world[0],
        p.camera_world[1] - p.planet_center_world[1],
        p.camera_world[2] - p.planet_center_world[2],
    ];

    // (id, score) in iteration order; stable-sorted later (matches JS stable sort).
    let mut candidates: Vec<(usize, f64)> = Vec::new();
    // insertion-ordered parent aggregation (children count, max area).
    let mut parent_order: Vec<usize> = Vec::new();
    let mut parent_children: std::collections::HashMap<usize, (u32, f64)> =
        std::collections::HashMap::new();

    let inv3 = 1.0 / 3.0;

    for leaf in leaves {
        let cx = (leaf.v0[0] + leaf.v1[0] + leaf.v2[0]) * inv3;
        let cy = (leaf.v0[1] + leaf.v1[1] + leaf.v2[1]) * inv3;
        let cz = (leaf.v0[2] + leaf.v1[2] + leaf.v2[2]) * inv3;
        let centroid_local = [cx, cy, cz];

        let rotated = transform_normal(centroid_local, &p.render_mat);
        let centroid_world = [
            p.planet_center_world[0] + rotated[0],
            p.planet_center_world[1] + rotated[1],
            p.planet_center_world[2] + rotated[2],
        ];

        let dist = MIN_DISTANCE.max(distance(p.camera_world, centroid_world));
        let area_world = triangle_area(leaf.v0, leaf.v1, leaf.v2);
        let projected_area_px2 = area_world * focal2 / (dist * dist);

        if projected_area_px2 >= split_threshold {
            let mut culled = p.cull_backface
                && triangle_backface(
                    leaf.v0,
                    leaf.v1,
                    leaf.v2,
                    &p.render_mat,
                    cam_rel,
                    p.cull_min_dot,
                );

            if !culled {
                if let Some(planes) = p.frustum_planes {
                    let mut br2 = 0.0;
                    for v in [leaf.v0, leaf.v1, leaf.v2] {
                        let dx = v[0] - cx;
                        let dy = v[1] - cy;
                        let dz = v[2] - cz;
                        let d2 = dx * dx + dy * dy + dz * dz;
                        if d2 > br2 {
                            br2 = d2;
                        }
                    }
                    let bound_radius = br2.sqrt();
                    let rel = [
                        centroid_world[0] - p.camera_world[0],
                        centroid_world[1] - p.camera_world[1],
                        centroid_world[2] - p.camera_world[2],
                    ];
                    culled = outside_frustum(rel, bound_radius, planes, p.frustum_guard_scale);
                }
            }

            if !culled {
                candidates.push((leaf.id, projected_area_px2));
            }
        }

        if leaf.parent >= 0 {
            let key = leaf.parent as usize;
            match parent_children.get_mut(&key) {
                Some(entry) => {
                    entry.0 += 1;
                    if projected_area_px2 > entry.1 {
                        entry.1 = projected_area_px2;
                    }
                }
                None => {
                    parent_order.push(key);
                    parent_children.insert(key, (1, projected_area_px2));
                }
            }
        }
    }

    // Stable sort desc by score (JS Array.sort is stable since ES2019).
    candidates.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap());
    let split_candidates: Vec<usize> = candidates.into_iter().map(|(id, _)| id).collect();

    // Filter eligible parents (2 leaf children, below merge threshold), then stable
    // sort asc by max child area — preserving first-seen order for ties.
    let mut merge: Vec<(usize, f64)> = parent_order
        .into_iter()
        .filter_map(|pid| {
            let (children, max_area) = parent_children[&pid];
            if children == 2 && max_area <= merge_threshold {
                Some((pid, max_area))
            } else {
                None
            }
        })
        .collect();
    merge.sort_by(|a, b| a.1.partial_cmp(&b.1).unwrap());
    let merge_parents: Vec<usize> = merge.into_iter().map(|(id, _)| id).collect();

    ClassifyResult {
        split_candidates,
        merge_parents,
    }
}
